new recipe: splunk-opentelemetry-cpp

seemk opened the pull request:
**splunk-opentelemetry-cpp/0.4.0**

Hi, this is a first to-be-published version of [Splunk's distribution of OpenTelemetry for C++](https://github.com/signalfx/splunk-otel-cpp). This won't be a dependency of other libraries we intend to package. I'm one of the devs for the distribution and the goal is to ease the usage process.

Any suggestions welcome, thanks :slightly_smiling_face: 

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=7565) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=7565) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=7565) it.</sub>

seemk commented:
It looks the includes are borked, fixing it

seemk commented:
@madebr @uilianries Updated/rewrote the recipe to use `opentelemetry/1.0.1` as a dependency. Waiting on https://github.com/conan-io/conan-center-index/pull/7799 now

seemk commented:
The Windows job seems to get stuck every time :thinking:

uilianries commented:
Let's restart.

seemk commented:
@uilianries Perhaps I should recreate this PR?

uilianries commented:
@seemk we have had an internal problem, open a new PR won't solve this situation. Conan team is working to fix it asap.

seemk commented:
> @seemk we have had an internal problem, open a new PR won't solve this situation. Conan team is working to fix it asap.

Thanks, appreciated!

seemk commented:
@uilianries Is there anything that I can do to help here?

madebr commented:
It looks like Windows is hanging.
Perhaps the test is failing there?
The Linux logs show:
```
----Running------
> bin/test_package
-----------------
[Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
```

seemk commented:
> It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> 
> ```
> ----Running------
> > bin/test_package
> -----------------
> [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> ```

Thanks, this message is expected though (default conf tries to export via OTLP) and the program exits normally. I can't see the Windows logs, so no idea what's going on there :slightly_frowning_face:

madebr commented:
Looks like it does not support dll's.
You can fix it in 2 ways: a dirty way or the proper way.
- Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
[The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
- Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.

prince-chrismc commented:
My vote is the dirty way :broom: :bucket:

seemk commented:
> Looks like it does not support dll's. You can fix it in 2 ways: a dirty way or the proper way.
> 
> * Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
>   It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
>   [The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
> * Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.

Thanks! I changed the library to use `GenerateExportHeader`.

> It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> 
> ```
> ----Running------
> > bin/test_package
> -----------------
> [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> ```

And you were right about this one, I think it's an issue with OpenTelemetry C++ on Windows where it's waiting indefinitely to flush the spans :thinking: and the export timeouts for OTLP exporters are not working (even though they are configurable). I've changed it to Jaeger exporter which does have proper timeouts.


Should be all good now, thanks for the help everyone. @madebr @prince-chrismc

madebr commented:
> > Looks like it does not support dll's. You can fix it in 2 ways: a dirty way or the proper way.
> > 
> > * Proper way: patch the header to add `__declspec(dllexport)`/`__declspec(dllimport)` to the correct spaces
> >   It looks like this is needed in 2 places: [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L26) and [here](https://github.com/signalfx/splunk-otel-cpp/blob/80be2accdf4720ce7f621503adfa0be04d22623b/include/splunk/opentelemetry.h#L45-L46).
> >   [The cmake `GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) might be useful for this.
> > * Dirty way: add `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` to the cmake wrapper.
> 
> Thanks! I changed the library to use `GenerateExportHeader`.

Great! Using this cmake module and not hand rolling this also makes it a more smooth experience for your users.
The downside is that usages of the same headers and switching between static/shared splunk-opentelemetry-cpp libraries is impossible now (as is the case on linux systems): the headers fix the library type.

> > It looks like Windows is hanging. Perhaps the test is failing there? The Linux logs show:
> > ```
> > ----Running------
> > > bin/test_package
> > -----------------
> > [Error] File: /home/conan/w/BuildSingleReference/.conan/data/opentelemetry-cpp/1.0.1/_/_/build/68e1659b012a357550b3f136fa4a0741e68f62d1/source_subfolder/exporters/otlp/src/otlp_grpc_exporter.cc:149[OTLP Exporter] Export() failed: failed to connect to all addresses
> > ```
> 
> And you were right about this one, I think it's an issue with OpenTelemetry C++ on Windows where it's waiting indefinitely to flush the spans thinking and the export timeouts for OTLP exporters are not working (even though they are configurable). I've changed it to Jaeger exporter which does have proper timeouts.
> 
> Should be all good now, thanks for the help everyone. @madebr @prince-chrismc

Running your library on a new platform often gives surprises and headaches.
I'm happy you were able to fix it and and up with a better library.

prince-chrismc commented on recipes/splunk-opentelemetry-cpp/all/conanfile.py:
```diff
@@ -0,0 +1,84 @@
+
+class SplunkOpentelemetryConan(ConanFile):
+    name = "splunk-opentelemetry-cpp"
+    license = "Apache 2.0"
```
https://spdx.org/licenses/Apache-2.0.html

```suggestion
    license = "Apache-2.0"
```

prince-chrismc commented on recipes/splunk-opentelemetry-cpp/all/conanfile.py:
```diff
@@ -0,0 +1,84 @@
+
+    def _remove_unnecessary_package_files(self):
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
There's no `pkg_config` target declared in `package_info`

seemk replied:
Ah thanks, the `pkgconfig` deletion was a remnant from an earlier recipe version. Removed it now

prince-chrismc reviewed (commented):
Super tiny comments. It looks really good!!!

seemk commented:
@uilianries Think this needs your powers here

prince-chrismc commented:
Ping since the PR is blocked

#7565 | seemk | Oct 6 | splunk-opentelemetry-cpp | 25 | Nov 24 | @uilianries | madebr, prince-chrismc, danimtb
-- | -- | -- | -- | -- | -- | -- | --

seemk commented:
@uilianries :pray:

prince-chrismc commented:
Today might be his first day back from vacation based on his status. Shouldn't be too long

prince-chrismc commented:
Ping @uilianries since the PR is blocked

uilianries reviewed (approved):
LGTM sorry my delay
